### DISCUSSION TITLE
Feature request: MotionCor3 integration

### DISCUSSION THREAD

**Discussion by alexjnoble** [nextpyp/.github#16] [Ideas]
Hello,

I think it would be nice for users to have access to multiple frame alignment methods. Could MotionCor3 be integrated?

https://github.com/czimaginginstitute/MotionCor3

MotionCor3 is BSD-3, so integration shouldn't be a problem. There is no Singularity build, so that would need to be created if necessary. There is also CTF estimation that I haven't tested.

Thank you,
-Alex

**Comment by asarnow** [comment]
I can offer to help with building MotionCor3 or explaining its options if that would be useful.

**Comment by abartesaghi** [comment]
Thanks @alexjnoble for the suggestion and @asarnow for offering to help. The same comments I made for [AreTomo2](https://github.com/orgs/nextpyp/discussions/15#discussioncomment-7746568) apply here, but I agree that giving users more options for frame-alignment will be beneficial. We will look into integrating MotionCor3 in nextPYP.

**Reply by abartesaghi** [reply]
Support for AreTomo2 and MotionCor3 is available starting with version [0.6.0](https://nextpyp.app/files/pyp/0.6.0/docs/index.html). Thanks to @alexjnoble and Jake Johnston for their input in setting default program options.

**Reply by xingzhe1129** [reply]
I am trying to run Motioncor3 with GPUs, my job commands are "/opt/pyp/bin/run/pyp" -scope_pixel 0.4125 -scope_voltage 300.0 -scope_dose_rate 60.0 -data_path "/lab-share/PCMM-Wu-e2/Public/xip/apol1/0902ap12022/Raw_data/*" -data_mode spr -gain_reference "/lab-share/PCMM-Wu-e2/Public/xip/apol1/0902ap12022/Raw_data/gain.mrc" -movie_ali motioncor -detect_rad 70.0 -detect_method all -detect_dist 30 -slurm_memory 5 -slurm_gres "--gres=gpu:NVIDIA_A100:1" -slurm_bundle_size 1000 -slurm_queue "bch-compute-pe" -slurm_queue_gpu "wu-gpu" -slurm_merge_memory 5 -slurm_merge_gres "--gres=gpu:NVIDIA_A100:1" -slurm_merge_queue "bch-compute-pe". But it always said "pyp: error: argument -slurm_gres/--slurm_gres: expected one argument" and job failed. Could you please help me figure it out ? Thanks.

**Reply by abartesaghi** [reply]
The new Gres options should not really be used to request GPU resources, that is handled directly by the website. The Gres options are meant to be used to request other types of resources such as local scratch, etc. Also, you shouldn't include the suffix `--gres=` when specifying these options, you only need to specify the resource itself, nextPYP will automatically add the suffix for you. The merge task is a lightweight job that doesn't use GPUs, so you don't need to request GPUs for merge tasks using the `Gres (merge task)` parameter. That being said, right now there really isn't a way to request a specific type of GPU, such as `gpu:NVIDIA_A100:1`. Whenever GPUS are needed, the website will use `--gres=gpu:1` and the type of GPU will be whatever is available in the gpu queue that was selected. Bottom line, try resetting all you Gres options and re-launching the job.

**Reply by xingzhe1129** [reply]
I reset Gres options as gpu:1, but it seems still doesn't work. Can I set -slurm_queue_gpu as bch-gpu, and set -slurm_queue as bch-compute, which means cpu and gpu partition are different nodes? Is this ok? Thanks.

**Reply by abartesaghi** [reply]
Right, since nextPYP is already including `gpu:1`, your Gres options need to be empty. Otherwise, `gpu:1` will be included twice and slurm will produce an error. Yes, it is OK to have two separate partitions. nextPYP will automatically submit any jobs that require GPUs to the gpu partition that was configured for your instance.

**Reply by xingzhe1129** [reply]
We can't change parameters for patch, bin, framedose, pixelsize and Sumrange in MotionCor3. Could you please try to add these options? Thanks.

![Screenshot 2024-01-24 at 14 46 10](https://github.com/nextpyp/.github/assets/76068275/a4157acf-cd35-4b46-802d-41197cd8c9ea)

**Comment by abartesaghi** [comment]
You can control the number of patches using the `Number of patches (patch tracking)` parameter from the `Tilt-series alignment` tab. We know this is counterintuitive, but at least it gives you a way of doing what you want. Now that we are supporting more external programs for executing various processing steps, we are working on refactoring the way parameters are entered to make it more intuitive.

nextPYP doesn't have an equivalent of the `-FmDose` parameter because the weighting is done based on the data itself during 3D refinement. Instead of using `-SumRange`, you may be able to achieve a similar effect using the `First frame` and `Last frame` options from the `Frame alignment` tab.

We definitely plan to add parameters like `-FtBin` and other motioncor3-specific options once the new interface to enter parameters is in place.

**Reply by xingzhe1129** [reply]
I am trying to do motion correction for single particle data set in nextPYP and can't control patches in tilt-series alignment tab.

**Reply by abartesaghi** [reply]
OK, in that case you won't be able to use patches in the current version. We will expose this and other motioncor3 parameters in the next maintenance release. Thanks for the feedback.

**Reply by xingzhe1129** [reply]
I set bch-compute as CPU partition for launch task and set bch-gpu as GPU partition for running Motioncor3, but the job was only ran on bch-compute cpu partition and can't run on bch-gpu gpu partition. Do you think it is the problem of my setting or a bug? Thanks.

**Reply by tofushan** [reply]
If you select `motioncor3` for movie frame alignment, it should use GPU for the entire pre-processing. Please check if you select `motioncor3` as your movie frame alignment option and set `GPU partition` to your gpu machines.

**Reply by xingzhe1129** [reply]
Yes, I do select motioncor3 and set gpu partition. But I can't set bch-gpu (gpu partition) as cpu partition for launch task. I guess the cpu partition for launch task should be the same node as gpu partition for running motioncor3. Is this correct? Thanks.

**Reply by tofushan** [reply]
To my knowledge, the resources for launching jobs and running jobs can be different, but only CPUs can be used to launch jobs. I'm adding @abartesaghi to confirm.

**Reply by abartesaghi** [reply]
This was an intermittent issue that should now be addressed in version [0.6.1](https://nextpyp.app/files/pyp/0.6.1/docs/index.html).

**Reply by xingzhe1129** [reply]
Thank you very much. The GPUs for motioncor3 work well now. But it seems that there is not full set of options when using MotionCor3 for frame alignment. binnning factor does't work for unblur and motioncor3, and images are still super-resolution size after motion correction. And patch for motioncor3 alignment should at least have 2 numbers and I don't know how 1 number works. Thanks for your reply.

**Reply by abartesaghi** [reply]
You are right, this was an oversight from our part. We will fix this in the upcoming minor release. The way it is setup now, the program uses the same number of patches in x and y, so if you set `Number of patches` to 5, it will call motioncor3 using `-Patch 5 5`.

**Reply by abartesaghi** [reply]
These issues are now addressed in version [0.6.2](https://nextpyp.app/files/pyp/0.6.2/docs/index.html).
